### PR TITLE
State: Introduce Jetpack user connection data to Redux

### PR DESCRIPTION
**Pull request by tyxla**
This PR introduces the Redux functionality for retrieving the connection data of the current Jetpack user. It will be necessary in order to create settings blocks, related with managing the user's connection. This PR lays most of the groundwork for #13232.

The PR introduces the following:

* `getJetpackUserConnectionData()` wpcom undocumented method to query the Jetpack site's REST API and fetch the user connection data.
* `requestJetpackUserConnectionData` Redux action to trigger a user connection data request.
* `dataItems` and `dataRequests` reducers to handle user connection data requests and items.
* `isRequestingJetpackUserConnection` selector that checks whether a user connection data request is in progress for a site.
* `getJetpackUserConnection` selector that retrieves the user connection data for a particular site.
* `isJetpackUserMaster` selector that returns whether the current user is the master user that owns the connection.
* `QueryJetpackUserConnection` query component to manage fetching of the user connection data.
* Tests for all of the above.

To test:

* Checkout this branch
* Verify all of the following tests pass:
  * `npm run test-client client/state/jetpack/connection/test/actions.js`
  * `npm run test-client client/state/jetpack/connection/test/reducer.js`
  * `npm run test-client client/state/selectors/`
* Drop the `QueryJetpackUserConnection` component somewhere and verify it performs the expected network requests.

### CODE REVIEW

File: client/state/jetpack/connection/reducer.js [hunk 1]
```diff
@@ -45,6 +49,32 @@ export const requests = createReducer( {}, {
...
+/**
+ * `Reducer` function which handles request/response actions
+ * concerning Jetpack user connection data
+ *
+ * @param  {Array}  state  Current state
+ * @param  {Object} action action
+ * @return {Array}         Updated state
+ */
+export const dataItems = createReducer( {}, {
+	[ JETPACK_USER_CONNECTION_DATA_RECEIVE ]: ( state, { siteId, data } ) => Object.assign( {}, state, { [ siteId ]: data } )
```
Comment (sirreal) [inline client/state/jetpack/connection/reducer.js]: It would be nice to use `keyedReducer` [like we do here](https://github.com/Automattic/wp-calypso/blob/2f73cb1c0342253c7cc442a271ea2951cd943b84/client/state/activity-log/restore/reducer.js#L43-L47).

File: client/state/jetpack/connection/reducer.js [hunk 2]
```diff
@@ -45,6 +49,32 @@ export const requests = createReducer( {}, {
...
+/**
+ * `Reducer` function which handles request/response actions
+ * concerning Jetpack user connection data requests
+ *
+ * @param {Object} state - current state
+ * @param {Object} action - action
+ * @return {Object} updated state
+ */
+export const dataRequests = createReducer( {}, {
+	[ JETPACK_USER_CONNECTION_DATA_REQUEST ]: createRequestReducer( true ),
```
Comment (sirreal) [inline client/state/jetpack/connection/reducer.js]: Again, it would be nice to use `keyedReducer` [like we do here](https://github.com/Automattic/wp-calypso/blob/2f73cb1c0342253c7cc442a271ea2951cd943b84/client/state/activity-log/restore/reducer.js#L43-L47).

File: client/lib/wpcom-undocumented/lib/undocumented.js [hunk 3]
```diff
@@ -305,6 +305,22 @@ Undocumented.prototype.getJetpackConnectionStatus = function( siteId, fn ) {
...
 };
 
+/*
+ * Retrieve current user's connection data for a Jetpack site.
+ *
+ * @param {int}      [siteId]
+ * @param {Function} fn
+ * @api public
+ */
+Undocumented.prototype.getJetpackUserConnectionData = function( siteId, fn ) {
```
Comment (sirreal) [inline client/lib/wpcom-undocumented/lib/undocumented.js]: Did you consider using the data-layer? One of the immediate benefits is to make `isRequesting` logic redundant.

File: client/components/data/query-jetpack-user-connection/index.jsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+class QueryJetpackUserConnection extends Component {
+	static propTypes = {
+		siteId: PropTypes.number,
+		// Connected props
+		isRequesting: PropTypes.bool,
+		requestJetpackUserConnectionData: PropTypes.func.isRequired,
+	}
+
+	componentDidMount() {
+		this.request( this.props );
```
Comment (oskosk) [inline client/components/data/query-jetpack-user-connection/index.jsx]: Can we make this query component be aware of `siteId === null` and avoid the request if that's the case ?

Comment (tyxla) [inline client/components/data/query-jetpack-user-connection/index.jsx]: We're already doing that, see https://github.com/Automattic/wp-calypso/pull/15599/files#diff-6a463ab01bf85f9ede30bd7a4663e845R33

File: client/lib/wpcom-undocumented/lib/undocumented.js (hunk 3 above)
Comment (tyxla) [inline client/lib/wpcom-undocumented/lib/undocumented.js]: Yes, that's a good point. I intentionally didn't use the data layer though, because I strived for consistency with the rest of the methods that communicate with the Jetpack site REST API. 

It would be a good improvement, of course, but I'd prefer to leave it for another PR (or set of PRs) that would migrate all of these actions to use the data layer.

File: client/components/data/query-jetpack-user-connection/index.jsx (hunk 4 above)
Comment (oskosk) [inline client/components/data/query-jetpack-user-connection/index.jsx]: Yes, but if the id is `0` the request won't ever be made. This is probably the id we would be using when trying Calypso in wp-admin for a single site installation.

File: client/lib/wpcom-undocumented/lib/undocumented.js (hunk 3 above)
Comment (oskosk) [inline client/lib/wpcom-undocumented/lib/undocumented.js]: Yes, the `isRequesting` logic is just redundant (in contrast to how the data layer handles tha tfor us) when used to check if a request is already being done from inside a Query component, but we're trusting in this flag for UI matters all along settings too.

File: client/components/data/query-jetpack-user-connection/index.jsx (hunk 4 above)
Comment (oskosk) [inline client/components/data/query-jetpack-user-connection/index.jsx]: My observation is probably outside of the scope of this PR tho. Having this PR take care of that id=`0`, wont' liberate us of the bulk of work needed to update all of the other query components :)

File: client/state/jetpack/connection/reducer.js (hunk 1 above)
Comment (tyxla) [inline client/state/jetpack/connection/reducer.js]: Thanks, I always forget about this one! Addressed in https://github.com/Automattic/wp-calypso/pull/15599/commits/6e87af6df408064765d44afdf4b8196ead451e7f.

File: client/state/jetpack/connection/reducer.js (hunk 2 above)
Comment (tyxla) [inline client/state/jetpack/connection/reducer.js]: Thanks, I always forget about this one! Addressed in https://github.com/Automattic/wp-calypso/pull/15599/commits/6e87af6df408064765d44afdf4b8196ead451e7f.

File: client/components/data/query-jetpack-user-connection/index.jsx (hunk 4 above)
Comment (tyxla) [inline client/components/data/query-jetpack-user-connection/index.jsx]: While we intend to be using `0` for wp-admin requests, it's still theoretical. And you're right we'll still have to update every other query component, so I suggest that we leave that for another PR.

**Comment by tyxla** [comment]
@oskosk @sirreal thanks for the feedback! I've addressed it and left comments on some of it, so this PR could use another 👁 . Thanks!

**Comment by oskosk** [comment]
This works as expected. Code looks solid and the query component worked for a master user and an admin user (not master). Tried the query component in settings/general for the master user and in sharing for an admin user (as we don't currently give them access to settings in Calypso)

🚢
